File: python3/Hash Maps and Sets/longest_chain_of_consecutive_numbers.py
from typing import List


def longest_chain_of_consecutive_numbers(nums: List[int]) -> int:
    if not nums:
        return 0
    num_set = set(nums)
    longest_chain = 0
    for num in num_set:
        # If the current number is the smallest number in its chain, search for
        # the length of its chain.
        if num - 1 not in num_set:
            current_num = num
            current_chain = 1
            # Continue to find the next consecutive numbers in the chain.
            while current_num + 1 in num_set:
                current_num += 1
                current_chain += 1
            longest_chain = max(longest_chain, current_chain)
    return longest_chain

File: cpp/Hash Maps and Sets/longest_chain_of_consecutive_numbers.cpp
#include <vector>        
#include <unordered_set> 
#include <algorithm>    

int longestChainOfConsecutiveNumbers(std::vector<int>& nums) {
    if (nums.empty()) {
        return 0;
    }
    std::unordered_set<int> numSet(nums.begin(), nums.end());
    int longestChain = 0;
    for (int num : numSet) {
        // If the current number is the smallest number in its chain, search for
        // the length of its chain.
        if (numSet.find(num - 1) == numSet.end()) {
            int currentNum = num;    
            int currentChain = 1;    
            // Continue to find the next consecutive numbers in the chain.
            while (numSet.find(currentNum + 1) != numSet.end()) {
                currentNum += 1;    
                currentChain += 1;   
            }
            longestChain = std::max(longestChain, currentChain);
        }
    }
    return longestChain;
}